### PR TITLE
Initial "Dotnet list package json output" implementation

### LINKED ISSUE #7752: Machine readable output for dotnet list package
## Details about Problem

It is difficult to parse the output of `dotnet list package`. We have requirements in our organization to audit packages that we are using in our projects. Without having a good way to parse the output of this command, it makes it difficult to structure the list of packages in a way that can be fed into another auditing system. 

Running the following command produces the following output:

```
dotnet list package --include-transitive
```

```
Project 'UniversalDashboard' has the following package references
   [net472]:
   Top-level Package                                           Requested          Resolved
   > Microsoft.AspNetCore                                      2.1.3              2.1.3
   > Microsoft.AspNetCore.Authentication.Cookies               2.1.2              2.1.2
   > Microsoft.AspNetCore.Authentication.Facebook              2.1.2              2.1.2
   > Microsoft.AspNetCore.Authentication.Google                2.1.2              2.1.2
   > Microsoft.AspNetCore.Authentication.JwtBearer             2.1.2              2.1.2
   > Microsoft.AspNetCore.Authentication.MicrosoftAccount      2.1.2              2.1.2
   > Microsoft.AspNetCore.Authentication.Twitter               2.1.2              2.1.2
   > Microsoft.AspNetCore.Mvc                                  2.1.2              2.1.2
   > Microsoft.AspNetCore.ResponseCompression                  2.1.1              2.1.1
   > Microsoft.AspNetCore.Routing                              2.1.1              2.1.1
   > Microsoft.AspNetCore.Server.Kestrel.Https                 2.1.3              2.1.3
   > Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv       2.1.3              2.1.3
   > Microsoft.AspNetCore.Session                              2.1.1              2.1.1
   > Microsoft.AspNetCore.SignalR                              1.0.3              1.0.3
   > Microsoft.AspNetCore.StaticFiles                          2.1.1              2.1.1
   > Microsoft.AspNetCore.WebSockets                           2.1.1              2.1.1
   > Microsoft.DotNet.PlatformAbstractions                     2.1.0              2.1.0
   > Microsoft.Extensions.Caching.Memory                       2.1.2              2.1.2
   > Microsoft.Extensions.Logging.Debug                        2.1.1              2.1.1
   > Microsoft.Extensions.Logging.Filter                       1.1.2              1.1.2
   > Newtonsoft.Json                                           11.0.2             11.0.2
   > NLog                                                      4.5.8              4.5.8
   > NLog.Web.AspNetCore                                       4.6.0              4.6.0
   > PowerShellStandard.Library                                5.1.0-preview-06   5.1.0-preview-06
   > Quartz                                                    3.0.6              3.0.6
   > System.Net.NameResolution                                 4.3.0              4.3.0
   > System.Reflection.Emit                                    4.3.0              4.3.0
   > System.Runtime.Loader                                     4.3.0              4.3.0
   > System.Security.Principal                                 4.3.0              4.3.0

   Transitive Package                                                Resolved
   > Libuv                                                           1.10.0
   > Microsoft.AspNetCore.Antiforgery                                2.1.1
   > Microsoft.AspNetCore.Authentication                             2.1.2
   > Microsoft.AspNetCore.Authentication.Abstractions                2.1.1
   > Microsoft.AspNetCore.Authentication.Core                        2.1.1
   > Microsoft.AspNetCore.Authentication.OAuth                       2.1.2
   > Microsoft.AspNetCore.Authorization                              2.1.1
   > Microsoft.AspNetCore.Authorization.Policy                       2.1.1
   > Microsoft.AspNetCore.Connections.Abstractions                   2.1.3
   > Microsoft.AspNetCore.Cors                                       2.1.1
   > Microsoft.AspNetCore.Cryptography.Internal                      2.1.1
   > Microsoft.AspNetCore.DataProtection                             2.1.1
   > Microsoft.AspNetCore.DataProtection.Abstractions                2.1.1
   > Microsoft.AspNetCore.Diagnostics                                2.1.1
   > Microsoft.AspNetCore.Diagnostics.Abstractions                   2.1.1
   > Microsoft.AspNetCore.HostFiltering                              2.1.1
   > Microsoft.AspNetCore.Hosting                                    2.1.1
   > Microsoft.AspNetCore.Hosting.Abstractions                       2.1.1
   > Microsoft.AspNetCore.Hosting.Server.Abstractions                2.1.1
   > Microsoft.AspNetCore.Html.Abstractions                          2.1.1
   > Microsoft.AspNetCore.Http                                       2.1.1
   > Microsoft.AspNetCore.Http.Abstractions                          2.1.1
   > Microsoft.AspNetCore.Http.Connections                           1.0.3
   > Microsoft.AspNetCore.Http.Connections.Common                    1.0.3
   > Microsoft.AspNetCore.Http.Extensions                            2.1.1
   > Microsoft.AspNetCore.Http.Features                              2.1.1
   > Microsoft.AspNetCore.HttpOverrides                              2.1.1
   > Microsoft.AspNetCore.JsonPatch                                  2.1.1
   > Microsoft.AspNetCore.Localization                               2.1.1
   > Microsoft.AspNetCore.Mvc.Abstractions                           2.1.2
   > Microsoft.AspNetCore.Mvc.ApiExplorer                            2.1.2
   > Microsoft.AspNetCore.Mvc.Core                                   2.1.2
   > Microsoft.AspNetCore.Mvc.Cors                                   2.1.2
   > Microsoft.AspNetCore.Mvc.DataAnnotations                        2.1.2
   > Microsoft.AspNetCore.Mvc.Formatters.Json                        2.1.2
   > Microsoft.AspNetCore.Mvc.Localization                           2.1.2
   > Microsoft.AspNetCore.Mvc.Razor                                  2.1.2
   > Microsoft.AspNetCore.Mvc.Razor.Extensions                       2.1.1
   > Microsoft.AspNetCore.Mvc.RazorPages                             2.1.2
   > Microsoft.AspNetCore.Mvc.TagHelpers                             2.1.2
   > Microsoft.AspNetCore.Mvc.ViewFeatures                           2.1.2
   > Microsoft.AspNetCore.Razor                                      2.1.1
   > Microsoft.AspNetCore.Razor.Design                               2.1.1
   > Microsoft.AspNetCore.Razor.Language                             2.1.1
   > Microsoft.AspNetCore.Razor.Runtime                              2.1.1
   > Microsoft.AspNetCore.ResponseCaching.Abstractions               2.1.1
   > Microsoft.AspNetCore.Routing.Abstractions                       2.1.1
   > Microsoft.AspNetCore.Server.IISIntegration                      2.1.1
   > Microsoft.AspNetCore.Server.Kestrel                             2.1.3
   > Microsoft.AspNetCore.Server.Kestrel.Core                        2.1.3
   > Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions      2.1.3
   > Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets           2.1.3
   > Microsoft.AspNetCore.SignalR.Common                             1.0.3
   > Microsoft.AspNetCore.SignalR.Core                               1.0.3
   > Microsoft.AspNetCore.SignalR.Protocols.Json                     1.0.3
   > Microsoft.AspNetCore.WebUtilities                               2.1.1
   > Microsoft.CodeAnalysis.Analyzers                                1.1.0
   > Microsoft.CodeAnalysis.Common                                   2.8.0
   > Microsoft.CodeAnalysis.CSharp                                   2.8.0
   > Microsoft.CodeAnalysis.Razor                                    2.1.1
   > Microsoft.CSharp                                                4.5.0
   > Microsoft.DiaSymReader.Native                                   1.7.0
```

The problem with this type of output is that it is difficult to parse. npm provides two options that make this easier when listing packages. They include both a `-parsable` and a `-json` parameter. I think `dotnet list package` should provide something similar. 

I think that `dotnet list package` should do something similar. 

```
F:\universal-dashboard\src\UniversalDashboard>dotnet --version
2.2.102
```

I'm happy to work on a PR for this. I've found the responsible code and could implement a "parse-friendly" version but wanted feedback from the community\team before doing anything. 

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/ProjectPackagesPrintUtility.cs

------------------------------
**Update 12/10/2021:**
After diving into this issue, it's quite large issue, it looks none of dotnet command implemented it, this is 1st time. Let's divide and conquer this issue, here're sub issues:
* Machine readable json output for dotnet list package: https://github.com/NuGet/Home/issues/11449
Spec proposal PR: https://github.com/NuGet/Home/pull/11446

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/NuGet/Home/issues/7752

Regression? Last working version:

## Description
Implement [Dotnet List Package Machine Readable Json Output](https://github.com/NuGet/Home/blob/861ebd714d6677d8b5837033b4951459a4b3370c/proposed/2022/DotnetListPackageMachineReadableJsonOutput.md) spec ([spec PR](https://github.com/NuGet/Home/pull/11983)), e.g. `dotnet list package --format json` will give you json output. 
If there any error with json generation then application'll return `non-zero` value.
This is **draft implementation** just to check if my technical implementation design going correct direction, if anyone has concern then better to address concern going too far into wrong direction, since it's 1st json output for dotnet so always room for learn and improve. Ideally, we should be MVC style pattern but there's tight coupling between business logic and console output from previous implementation, it's very hard to make MVC style at this moment.
Please just quickly glance at PR and let me know if there are any serious concern, I know implementation is not complete there is some discrepancy what you see here and spec, I'll address them later.

` dotnet list package`
```
Project 'MyProjectA' has the following package references
   [netcoreapp3.1]:
   Top-level Package                      Requested             Resolved
   > DotNetNuke.Core                      6.0.0                 6.0.0
   > DotNetZip                            1.0.0                 1.0.0
   > EntityFramework.MappingAPI           *                     6.2.1
   > Microsoft.Extensions.Primitives      [1.0.0, 5.0.0]        1.0.0
   > NuGet.Core                           2.13.0                2.13.0
   > System.CommandLine                   0.3.0-alpha.20070.2   0.3.0-alpha.20070.2
   > Text2Xml.Lib                         [1.1.2, 2.0.0)        1.1.2

Project 'MyProjectB' has the following package references
   [netcoreapp3.1]:
   Top-level Package      Requested             Resolved
   > DotNetNuke.Core      6.0.0                 6.0.0
   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
   > NuGet.Core           2.13.0                2.13.0
   > Text2Xml.Lib         1.1.2                 1.1.2

   [net5.0]:
   Top-level Package      Requested             Resolved
   > DotNetNuke.Core      6.0.0                 6.0.0
   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
   > NuGet.Core           2.13.0                2.13.0
   > Text2Xml.Lib         1.1.2                 1.1.2
```

Here is current basic output looks like for `dotnet list package --format json`:
```
{
  "version": 1,
  "parameters": "",
  "projects": [
    {
      "path": "MyProjectA",
      "frameworks": [
        {
          "framework": "netcoreapp3.1",
          "topLevelPackages": [
            {
              "id": "DotNetNuke.Core",
              "requestedVersion": "6.0.0",
              "resolvedVersion": "6.0.0"
            },
            {
              "id": "DotNetZip",
              "requestedVersion": "1.0.0",
              "resolvedVersion": "1.0.0"
            },
            {
              "id": "EntityFramework.MappingAPI",
              "requestedVersion": "*",
              "resolvedVersion": "6.2.1"
            },
            {
              "id": "Microsoft.Extensions.Primitives",
              "requestedVersion": "[1.0.0, 5.0.0]",
              "resolvedVersion": "1.0.0"
            },
            {
              "id": "NuGet.Core",
              "requestedVersion": "2.13.0",
              "resolvedVersion": "2.13.0"
            },
            {
              "id": "System.CommandLine",
              "requestedVersion": "0.3.0-alpha.20070.2",
              "resolvedVersion": "0.3.0-alpha.20070.2"
            },
            {
              "id": "Text2Xml.Lib",
              "requestedVersion": "[1.1.2, 2.0.0)",
              "resolvedVersion": "1.1.2"
            }
          ]
        }
      ]
    },
    {
      "path": "MyProjectB",
      "frameworks": [
        {
          "framework": "netcoreapp3.1",
          "topLevelPackages": [
            {
              "id": "DotNetNuke.Core",
              "requestedVersion": "6.0.0",
              "resolvedVersion": "6.0.0"
            },
            {
              "id": "NuGet.Commands",
              "requestedVersion": "4.8.0-preview3.5278",
              "resolvedVersion": "4.8.0-preview3.5278"
            },
            {
              "id": "NuGet.Core",
              "requestedVersion": "2.13.0",
              "resolvedVersion": "2.13.0"
            },
            {
              "id": "Text2Xml.Lib",
              "requestedVersion": "1.1.2",
              "resolvedVersion": "1.1.2"
            }
          ]
        },
        {
          "framework": "net5.0",
          "topLevelPackages": [
            {
              "id": "DotNetNuke.Core",
              "requestedVersion": "6.0.0",
              "resolvedVersion": "6.0.0"
            },
            {
              "id": "NuGet.Commands",
              "requestedVersion": "4.8.0-preview3.5278",
              "resolvedVersion": "4.8.0-preview3.5278"
            },
            {
              "id": "NuGet.Core",
              "requestedVersion": "2.13.0",
              "resolvedVersion": "2.13.0"
            },
            {
              "id": "Text2Xml.Lib",
              "requestedVersion": "1.1.2",
              "resolvedVersion": "1.1.2"
            }
          ]
        }
      ]
    }
  ]
}
```
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added, I'll add tests after getting confirmation on my current implementation is good enough.
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Hi team.
Could take quick look at PR and check if my implementation is reasonable? If it looks good, then I'll continue the implementation.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputFormat.cs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+                if (reportProject.FrameworkPackages.Count > 0)
+                {
+                    writer.WritePropertyName(FrameworksProperty);
+
+                    WriteFrameworkPackage(writer, reportProject.FrameworkPackages);
+
+                }
+                writer.WriteEndObject();
+
+                //writer.WritePropertyName(IdProperty);
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputFormat.cs:102]: I'll clean up comments soon.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs [hunk 2]
```diff
@@ -158,8 +162,19 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
             // Print a legend message for auto-reference markers used
             if (autoReferenceFound)
             {
-                Console.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+                reportRenderer.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+
+                // Should log into json output ?
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:165]: `autoReferenceFound` is for the package referenced from sdk.
It's something we didn't consider for spec, so I far I haven't seen actual example of this.

Comment (zivkan) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:165]: Test a project that targets netstandard2.0. It should have an implicit package reference.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs [hunk 3]
```diff
@@ -50,10 +54,10 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
...
             if (listPackageArgs.ReportType != ReportType.Default)
             {
-                Console.WriteLine();
-                Console.WriteLine(Strings.ListPkg_SourcesUsedDescription);
-                ProjectPackagesPrintUtility.PrintSources(listPackageArgs.PackageSources);
-                Console.WriteLine();
+                reportRenderer.WriteLine();
+                reportRenderer.WriteLine(Strings.ListPkg_SourcesUsedDescription);
+                ProjectPackagesPrintUtility.PrintSources(listPackageArgs);
+                reportRenderer.WriteLine();
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: I think that this, coupled with the fact that `JsonRenderer` has a bunch of methods that are no-ops, the text renderer also has some methods that are no-ops, and the fact that there's a `if (render is JsonRenderer)` below, that there isn't a good match in abstractions.

Effectively, this `ListPackageCommandRunner` is written in a way that orders its work based on the console output, and then you're trying to bolt a json renderer on top of this.

Instead, I propose something similar to a web MVC architecture. A "controller" which gets the information for a single project (the model), and then the view (renderer) knows how to render one of these models. Something like:

```cs
renderer.Start(); // no-op for command line, for json it writes `{` and version property
foreach (var project in allProjects)
{
  // args is set up before the loop, from the command line arguments. It tells the controller
  // what properties of the model to populate, which is how it will know to be "offline only"
  // for certain text output scenarios.
  var model = await controller.GetPackagesAsync(project, args);
  renderer.Write(project);
}
renderer.End(); // json writes errors, other info?.   does console output any "wrapping up" info?
```

Honestly, the "model" could actually be the model for JSON output, and then the renderer is little more than a `JsonSerializer.Serialize(console.out, model)`, and the only place "logic" is needed is the text renderer.

This way `IReportRenderer` doesn't need a bunch of methods that are only implemented in one of the two renderers.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: > I think that this, coupled with the fact that `JsonRenderer` has a bunch of methods that are no-ops, the text renderer also has some methods that are no-ops, and the fact that there's a `if (render is JsonRenderer)` below, that there isn't a good match in abstractions.
> 

Removed `if (render is JsonRenderer)`  logic, it can be abstract away.

> Effectively, this `ListPackageCommandRunner` is written in a way that orders its work based on the console output, and then you're trying to bolt a json renderer on top of this.
> 
> Instead, I propose something similar to a web MVC architecture. A "controller" which gets the information for a single project (the model), and then the view (renderer) knows how to render one of these models. Something like:
> 
> ```cs
> renderer.Start(); // no-op for command line, for json it writes `{` and version property
> foreach (var project in allProjects)
> {
>   // args is set up before the loop, from the command line arguments. It tells the controller
>   // what properties of the model to populate, which is how it will know to be "offline only"
>   // for certain text output scenarios.
>   var model = await controller.GetPackagesAsync(project, args);
>   renderer.Write(project);
> }
> renderer.End(); // json writes errors, other info?.   does console output any "wrapping up" info?
> ```
> 
> Honestly, the "model" could actually be the model for JSON output, and then the renderer is little more than a `JsonSerializer.Serialize(console.out, model)`, and the only place "logic" is needed is the text renderer.
> 
> This way `IReportRenderer` doesn't need a bunch of methods that are only implemented in one of the two renderers.

Thank you for feedback. 
1. With this approach we couldn't print all the problems on top of json report, instead it would be written to the bottom of report. Is it ok?
Current json template is 
```
{
 "version": 1,
  "parameters": "",
  "problems": [
    {
        "project": "src/lib/MyProjectB.csproj",
        "message": "The project `C:/Users/userA/repos/MainApp/src/lib/MyProjectB.csproj`` uses package.config for NuGet packages, while the command works only with package reference projects."
    }
  ],
  "projects": [
    {
      ...
    }
  ]
}
```

Instead, we would do:
```

{
    "version": 1,
     "parameters": "",
         "projects": [
       {
         ...
       }
     ],
     "problems": [
        {
            "project": "src/lib/MyProjectB.csproj",
            "message": "The project `C:/Users/userA/repos/MainApp/src/lib/MyProjectB.csproj`` uses package.config for NuGet packages, while the command works only with package reference projects."
        }
      ]     
}

```
2. Also making it MVC style means I need to completely rewrite the current console output logic too. I'm not sure unless I divide into phases not sure how much additional effort it would take.

Comment (zivkan) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: > With this approach we couldn't print all the problems on top of json report, instead it would be written to the bottom of report. Is it ok?

Two things:

1. Section 1 of the [JSON specification](https://datatracker.ietf.org/doc/html/rfc7159#section-1) says:

    >    An object is an unordered collection of zero or more name/value
    >   pairs, where a name is a string and a value is a string, number,
    >   boolean, null, object, or array.
    >
    >   An array is an ordered sequence of zero or more values.

    In other words, it is "invalid" to depend on specific order of properties in an object.  Having said that, we really, really need version to be the very first property so that tools can implement polymorphic deserialization without buffering the entire json (for when a version 2 output eventually exists).

2. The JSON renderer's implementation of `void Write(ProjectInfo)` could be just to add the object into a list, and it doesn't actually output anything until the `void End()` at the end of the loop.

The big picture I was trying to raise awareness over is overall readability and "software architecture". Let's avoid leaky abstractions. Don't get bogged down in the example comments that were designed to help understand the API design. How different renderers implement the APIs are implementation specific.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputContent.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace NuGet.CommandLine.XPlat.ReportRenderers.JsonRenderer
+{
+    internal class JsonOutputContent
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputContent.cs]: Why don't just create an instance of this object and serialize it to json using a well-known json library (System.Text.Json or Newtonsoft.Json) ?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: > 
> 1. Section 1 of the [JSON specification](https://datatracker.ietf.org/doc/html/rfc7159#section-1) says:
>    > An object is an unordered collection of zero or more name/value
>    > pairs, where a name is a string and a value is a string, number,
>    > boolean, null, object, or array.
>    > An array is an ordered sequence of zero or more values.
>    
>    
>    In other words, it is "invalid" to depend on specific order of properties in an object.  Having said that, we really, really need version to be the very first property so that tools can implement polymorphic deserialization without buffering the entire json (for when a version 2 output eventually exists).

Of course, it doesn't make any difference for machine which order is objects are in json, but someone looking at the report have to scroll down all the way to find out what was the issue (for example admin looking at non-zero exit code from json output, in case there're hundreds of projects and thousands of dependencies then finding what went wrong requires writing own parse). Current structure allows to find out if there was issue at 1st glance and it's consistent what `npm -ls json` doing it, I feel having errors/problems at top is more user friendly.

> 2. The JSON renderer's implementation of `void Write(ProjectInfo)` could be just to add the object into a list, and it doesn't actually output anything until the `void End()` at the end of the loop.
> 

I added `void Write(ReportProject reportProject)` method into `IRenderer interface` and `Void End()` is doing the action rendering, maybe this could be not enough to address your concern, let me know if I need to make more drastic change.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputContent.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/JsonRenderer/JsonOutputContent.cs]: Sure, we can tweak serializer (with JsonConverter) more later but I'm more worried about main component of new design `IRenderer.cs` and `ConsoleWriter.cs`, `ListPackageJsonRenderer.cs`, their usage which is hard to change later. Please take a look at current design.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs (hunk 3 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: > but someone looking at the report have to scroll down all the way to find out what was the issue

The feature is called "machine readable".  You're right that this doesn't mean we should make it difficult for humans to read.  But customers are expected to use it in scripts or pipe to other commands, not to output to the console and look at manually.  Those scripts they create can reply the errors to their CI logs, to the console, to whereever. Alternatively, if the customer runs the command without `--format json`, we expect them to see the same errors (but in a more human friendly text output).

>  I feel having errors/problems at top is more user friendly.

Only if the output is piped to something like `more` or `less`, right?  If you just run `dotnet list pacakges --format json` and there are pages and pages of output, when the terminal stops scrolling, you're only going to see what's at the end, not at the beginning.

> I added void Write(ReportProject reportProject) method into IRenderer interface and Void End() is doing the action rendering, maybe this could be not enough to address your concern, let me know if I need to make more drastic change.

I feel that since both the text renderer and json renderer have interface methods which they ignore, it's a code smell that it's not well designed.  I get that it's quicker because you don't need to de-tangle business logic from presentation. But think about it from the point of view of someone who hasn't joined the NuGet team yet (or who doesn't review this PR). In the future they'll get some work item to work on `dotnet list package`, they'll be reading the code for the first time, and trying to understand why it's calling `renderer.WriteLine(someString)` followed by `renderer.Write(ProjectInfo)`. Why doesn't `renderer.Write(ProjectInfo` generate `someString` itself?  Another also, I know this isn't a performance hot path, but all those times where the business logic does `renderer.WriteLine(string.Format(...))`, the runtime will need to parse the format string, create a string builder, create the string, return to caller, which then gets passed to the json renderer, just to ignore it.  Not a performance hot path, but every code path is an opportunity to practise best practices.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/IReportRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal interface IReportRenderer
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/IReportRenderer.cs:8]: Suggestion: Make the IReportRendered independent of exit code

```mermaid
classDiagram
class IReportRenderer~T~:::abstract {
TextWriter writer
WriteOutput(T reportContent)
}
IReportRenderer~T~ <|-- ConsoleWriter : implements
IReportRenderer~T~ <|-- JsonPackageRendered : implements
```

Maybe rename `ConsoleWriter`, `JsonPackageRendered ` types to reflect that they implement `IReportRenderer`

I think exit code is independent of text output.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs [hunk 6]
```diff
@@ -157,12 +179,52 @@ private static ReportType GetReportType(bool isDeprecated, bool isOutdated, bool
...
+            if (outputFormat == ReportOutputFormat.Console)
+            {
+                return (ConsoleWriter.Instance, ReportOutputFormat.Console);
+            }
+
+            // currently only version 1 is available, so default to latest available version 1.
+            IReportRenderer jsonReportRenderer;
+
+            var currentlySupportedReportVersions = new List<string> { "1" };
+            // If customer pass unsupported version then default to latest available version and warn about unsupported version.
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs]: The comment here is warn, but we seem to log jsonReportRenderer.WriteErrorLine. 

Shouldn't this be a 1 exit code, invalid input? 
Seems better than ignoring a bad parameter.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:58]: > I feel that since both the text renderer and json renderer have interface methods which they ignore, it's a code smell that it's not well designed. 

I agree with this. 

The abstraction is there in order to facilitate fewer code changes which isn't good design. 

Part of the reason why we've always had trouble with making changes to dotnet list package has been the lack of proper abstractions. I feel like this exacerbates that. 

In other words, there's an interface with 10 methods, but only 5 of those methods are actually common.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs [hunk 7]
```diff
@@ -157,12 +179,52 @@ private static ReportType GetReportType(bool isDeprecated, bool isOutdated, bool
...
         }
 
+        private static (IReportRenderer, ReportOutputFormat) GetOutputType(string outputFormatOption, string outputVersionOption)
+        {
+            ReportOutputFormat outputFormat = ReportOutputFormat.Console;
+            if (!string.IsNullOrEmpty(outputFormatOption))
+            {
+                try
+                {
+                    outputFormat = EnumExtensions.GetValueFromName<ReportOutputFormat>(outputFormatOption);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:189]: why not use enum parse?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs [hunk 8]
```diff
@@ -157,12 +179,52 @@ private static ReportType GetReportType(bool isDeprecated, bool isOutdated, bool
...
+                }
+                catch (ArgumentException)
+                {
+                    string currentlySupportedFormat = "console, json";
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:193]: This should be generated by looking at all the values of the enum, not by hardcoding it.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs [hunk 9]
```diff
@@ -158,8 +160,13 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
...
             if (autoReferenceFound)
             {
-                Console.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+                reportRenderer.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+
+                // Should log into json output ?
             }
+
+            listPackageArgs.Renderer.End();
+            return listPackageArgs.Renderer.ExitCode();
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs:169]: The renderer having the exit code feels weird. 

Roughly, the design should be something that returns a model and then something else that renders that. 

The model should contain the failure information.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/IReportRenderer.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/ReportRenderers/IReportRenderer.cs:8]: I left the same feedback as @dominoFire in a separate comment.

**Review by nkolev92 — COMMENTED** [review]
Big issue why you're having a tough time writing easy to follow code is the lack of good abstractions in this implementation. 

I feel like this exacerbates that. 
dotnet list package is a command that's had a lot of changes/additions. I think we should take the time to get it done correctly so that we don't have these issues in the future. 

Fortunately, not of these types are public APIs, so we can do better.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs (hunk 8 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:193]: This one is temporary; I want feedback on main design.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs (hunk 7 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs:189]: image
As you can see enum parsing is case sensitive. Another way of dealing with this one is make upper case the `outputFormatOption` and also create enum all Upper case too.
